### ISSUE TITLE
Add CONFIG_MODVERSIONS symbol versioning blurb to patch author guide

### ISSUE THREAD

**Issue by joe-lawrence** [dynup/kpatch#1105]
_Note: We have a card on the trello board to implement an automated check, but I also had this brain-dump to consider for inclusion in the patch author guide as a section or separate link.  Once we have the automated check, we can add its text somewhere here.  Since it seems like it is easier to quote and make suggestions here than in markdown files, here it is mostly unformatted for now.  Once we settle on text, I can add markdown to the guide and send out for PR._

Symbol versioning

Background

CONFIG_MODVERSIONS enables an ABI check between exported kernel symbols and modules referencing those symbols, enforced on module load.  When building the kernel, preprocessor output from gcc -E for each source file is passed to scripts/genksyms.  The genksyms script recursively expands each exported symbol to its basic types.  A hash is generated for each symbol as it traverses back up the symbol tree.  The end result is a CRC for each exported function in Module.symvers file and embedded in the vmlinux kernel object itself.  

A similar checksumming is performed when building modules: referenced exported symbol CRCs are stored in the module’s __versions section (you can also find these in plain-text intermediate *.mod.c files.)

When the kernel loads a module, the symbol CRCs found in its __versions are compared to those of the kernel, if the two do not match, the kernel will refuse to load it:
```
<module>: disagrees about version of symbol <symbol>
<module>: Unknown symbol <symbol> (err -22)
```
Kpatch implications

It is rare, but possible for a kpatch to introduce inadvertent symbol CRC changes that are not true ABI changes.  The following conditions must occur:

1. The kpatch must modify the definition of an exported symbol.  For example, introducing a new header file may further define an opaque data type:  Before the kpatch, compilation unit U from the original kernel build only knew about a struct S declaration (not its complete type).  At the same time, U contains function F, which has an interface that references S.  If the kpatch adds a header file to U that now fully defines struct S { int a, b, c; }, its symbol type graph changes, CRCs generated for F are updated, but its ABI remains consistent.

2. The kpatch must introduce either a change or reference to F such that it is included in the resulting kpatch module.  This will force a __version entry based on the new CRC.

Kpatch avoidance

Kpatches should introduce new include files sparingly.  Whenever possible, extract only the additional definitions necessary to build the kpatch.

If additional header files or symbol definitions cannot be avoided, consider surrounding the offending include/definitions in an `#ifndef __GENKSYMS__` macro.  The genksyms script will skip over those blocks when performing its CRC calculations on the source.

But what about a real ABI change?

If a kpatch introduces a true ABI change, each of calling functions would consequently need to be updated in the kpatch module.  For unexported functions, this may be handled safely if the kpatch does indeed update all callers.  However, since motivation behind CONFIG_MODVERSIONS is to provide basic ABI verification between the kernel and modules for  **exported** functions, kpatch cannot safely change this ABI without worrying about breaking other out-of-tree drivers.  Those drivers have been built against the reference kernel's original set of CRCs and expect the original ABI.

To track down specifically what provoked a symbol CRC change, tools like kabi-dw can be employed to produce a richer report of symbol definition deltas.  For a kpatch-build, kabi-dw can be modified to operate on .o object files (not just .ko and vmlinux files) and the $CACHEDIR/tmp/{orig, patched} directories compared.

**Comment by jpoimboe** [comment]
> But what about a real ABI change?
> 
> If a kpatch introduces a true ABI change to function F, each of its callers will be consequently updated and included in the kpatch module.
> 
> To track down specifically what provoked a symbol CRC change, tools like kabi-dw can be employed to produce a richer report of symbol definition deltas. For a kpatch-build, kabi-dw can be modified to operate on .o object files (not just .ko and vmlinux files) and the $CACHEDIR/tmp/{orig, patched} directories compared.

Excellent writeup!  I'm thinking this part might need a little more clarification.  CRCs are only checked for exported symbols.  A real ABI change of an exported symbol could be a very bad idea if there's the possibility of third party modules.  And it's probably best to avoid it anyway because then you have to patch all the in-tree modules which use it.

**Comment by joe-lawrence** [comment]
> > But what about a real ABI change?
> > If a kpatch introduces a true ABI change to function F, each of its callers will be consequently updated and included in the kpatch module.
> > To track down specifically what provoked a symbol CRC change, tools like kabi-dw can be employed to produce a richer report of symbol definition deltas. For a kpatch-build, kabi-dw can be modified to operate on .o object files (not just .ko and vmlinux files) and the $CACHEDIR/tmp/{orig, patched} directories compared.
> 
> Excellent writeup! I'm thinking this part might need a little more clarification. CRCs are only checked for exported symbols. A real ABI change of an exported symbol could be a very bad idea if there's the possibility of third party modules. And it's probably best to avoid it anyway because then you have to patch all the in-tree modules which use it.

Good point about exported symbols, I'll add mention of that to the list of conditions above.  As far as clarification goes, are you talking about kabi-dw verification?  Or further warning about introducing a "true" ABI change to an exported function implementation?

**Comment by jpoimboe** [comment]
The latter.

**Comment by joe-lawrence** [comment]
Ok, I edited the comment, qualifying that the check is for exported symbols and then adding a bit more to the real ABI change part at the bottom.

**Comment by jpoimboe** [comment]
Sounds good.  Typo: "the since"

**Comment by joe-lawrence** [comment]
Created some markdown:
* [link to branch](https://github.com/joe-lawrence/kpatch/tree/author-guide-MODVERSIONS)
* [link to github markdown rendering](https://github.com/joe-lawrence/kpatch/blob/author-guide-MODVERSIONS/doc/patch-author-guide.md#exported-symbol-versioning)

@julien-thierry : I think you were working on kpatch-build detection of this scenario. Do you want to integrate this with your eventual pull request (so you can update with any user-visible messaging)?

**Comment by julien-thierry** [comment]
Sorry I missed this message.

I picked up the commit. However there no Signed-off-by tag. Can I add your name for the tag, given that you are the author?

**Comment by joe-lawrence** [comment]
> I picked up the commit. However there no Signed-off-by tag. Can I add your name for the tag, given that you are the author?

I can give it another once-over today with a fresh set of eyes to cleanup any awkward wording and add my signed-off-by tag if that works for you. Thanks!

**Comment by julien-thierry** [comment]
Of course that's good for me.

I'll pick that up when you're done. Thank you!

**Comment by joe-lawrence** [comment]
@julien-thierry : ok, I reworded a few things and provided a space for the detection work.  Also a proper commit msg and s-o-b line.  Feel free to update to match your patchset accordingly.

**Comment by joe-lawrence** [comment]
Fixed by PR #1114 ("Fix symvers")
